libraries/AP_Beacon/AP_Beacon_Marvelmind.h

```diff
@@ -42,63 +67,58 @@ class AP_Beacon_Marvelmind : public AP_Beacon_Backend
+        PositionValue * position_buffer;  // buffer of measurements
+        StationaryBeaconsPositions positions_beacons;
         bool verbose;   // verbose flag which activate console output, default: False
         bool pause;     //  pause flag. If True, class would not read serial data
         bool terminationRequired;   //  If True, thread would exit from main loop and stop
-        void (*receive_data_callback)(struct PositionValue position); //  receive_data_callback is callback function to recieve data
+        void (*receive_data_callback)(PositionValue position); //  receive_data_callback is callback function to receive data
 
-        // private variables
         uint8_t _last_values_count;
         uint8_t _last_values_next;
         bool _have_new_values;
     };
 
-    // constructor
-    AP_Beacon_Marvelmind(AP_Beacon &frontend, AP_SerialManager &serial_manager);
-
-    // return true if sensor is basically healthy (we are receiving data)
-    bool healthy();
-
-    // update
-    void update();
-
-private:
     enum {
         RECV_HDR,
         RECV_DGRAM
     } parse_state; // current state of receive data
 
-    struct MarvelmindHedge *hedge;
-    struct PositionValue cur_position;
-    struct StationaryBeaconPosition cur_beacon;
+    MarvelmindHedge *hedge;
+    PositionValue cur_position;
     uint8_t input_buffer[AP_BEACON_MARVELMIND_BUF_SIZE];
     uint16_t num_bytes_in_block_received;
     uint16_t data_id;
 
-    struct MarvelmindHedge* m_MarvelmindHedge;
     uint16_t calc_crc_modbus(uint8_t *buf, uint16_t len);
-    bool get_or_alloc_beacon(struct StationaryBeaconPosition &b, uint8_t address);
+    StationaryBeaconPosition* get_or_alloc_beacon(uint8_t address);
     uint8_t mark_position_ready();
-    void process_beacons_positions_datagram(struct StationaryBeaconPosition &b);
-    void process_beacons_positions_highres_datagram(struct StationaryBeaconPosition &b);
-    void process_position_highres_datagram(struct PositionValue &p);
-    void process_position_datagram(struct PositionValue &p);
+    void process_beacons_positions_datagram();
+    void process_beacons_positions_highres_datagram();
+    void process_position_highres_datagram(PositionValue &p);
+    void process_position_datagram(PositionValue &p);
     void create_marvelmind_hedge();
     void start_marvelmind_hedge();
-    bool get_position_from_marvelmind_hedge(struct PositionValue *position);
-    void stop_marvelmind_hedge();
+    void set_stationary_beacons_positions_and_distances();
 
+    //Variables for Ardupilot
     AP_HAL::UARTDriver *uart = nullptr;
     uint32_t last_update_ms = 0;
+
+    // cache the vehicle position in NED coordinates [m]
+    Vector3f vehicle_position_NED__m;
+    bool vehicle_position_initialized = false;
+
+    // cache the beacon positions in NED coordinates [m]
+    Vector3f beacon_position_NED__m[AP_BEACON_MARVELMIND_MAX_STATIONARY_BEACONS];
+    bool beacon_position_initialized = false;
```

Comment:
Static objects are put in the BSS section which is zero-initialized and dynamically objects are allocated with our _new_ overload that ensures zero initilization: https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Common/c++.cpp